File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word)
    {
        unordered_map<char, int> m;
        for(char c : word)
            m[c]++;
        
        unordered_map<char, int> m2;
        for(vector<char>& bo : board)
            for(char c : bo)
                m2[c]++;
        
        for(auto pair : m)
            if(m[pair.first] > m2[pair.first])
                return false;
        
        for(int i = 0; i < board.size(); i++)
            for(int j = 0; j < board[i].size(); j++)
                if(dfs(board, i, j, word, 0))
                    return true;
        return false;
    }
    
    bool dfs(vector<vector<char>>& board, int x, int y, string word, int p)
    {
        if(p == word.size())
            return true;
        
        if(x < 0 || y < 0 || x >= board.size() || y >= board[0].size() || board[x][y] != word[p])
            return false;
        
        char c = board[x][y];
        board[x][y] = '#';
        
        if(dfs(board, x - 1, y, word, p + 1))
            return true;
        
        if(dfs(board, x + 1, y, word, p + 1))
            return true;
        
        if(dfs(board, x, y - 1, word, p + 1))
            return true;
        
        if(dfs(board, x, y + 1, word, p + 1))
            return true;
        
        board[x][y] = c;
        
        return false;
    }
};

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        
        wordDict = Counter(word)
        boardDict = defaultdict(int)
        for a in board:
            for x in a:
                boardDict[x] += 1
        for l in wordDict:
            if not l in boardDict or boardDict[l] < wordDict[l]:
                return False
        
        def finder(x: int, y: int, c: int) -> bool:
            nonlocal board, word
            if c == len(word):
                return True
            if x < 0 or y < 0 or x >= len(board) or y >= len(board[0]) or board[x][y] != word[c]:
                return False
            a = board[x][y]
            board[x][y] = "-"
            u, l, r, d = finder(x - 1, y, c + 1), finder(x, y - 1, c + 1), finder(x, y + 1, c + 1), finder(x + 1, y, c + 1)
            board[x][y] = a
            return u or l or r or d
            
        for i in range(len(board)):
            for j in range(len(board[i])):
                if finder(i, j, 0):
                    return True
        return False